Support Python 3.14

Just adding version 3.14 to the testing matrix does not work. Some of our tests fail and it's not instantly obvious why. I suspect the fact that the "subprocess" module has changed default engine with this version may be the reason.

Rather than try to fix this while we drag around a mountain of technical debt it would be best to postpone it for a while. It still needs doing though.

But this patch breaks python 3.10,the metrics causing
```
E       FileNotFoundError: [Errno 2] No such file or directory: '.../pytest-19913/test_zero_config0/fab-workspace/zero-config/metrics/zero-config/metrics.json
```
aborts :(

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -27,7 +27,7 @@ jobs:
     strategy:
       max-parallel: 4
       matrix:
-        python-version: ['3.10', '3.11', '3.12', '3.13']
+        python-version: ['3.10', '3.11', '3.12', '3.13', '3.14']
 
     steps:
       # This step does not need a GA cache

--- a/source/fab/build_config.py
+++ b/source/fab/build_config.py
@@ -10,7 +10,6 @@ Contains the :class:`~fab.build_config.BuildConfig` and helper classes.
 import getpass
 import logging
 import os
-import sys
 import warnings
 from datetime import datetime
 from fnmatch import fnmatch
@@ -129,7 +128,7 @@ class BuildConfig():
         # turn off multiprocessing when debugging
         # todo: turn off multiprocessing when running tests, as a good test
         # runner will run using mp
-        if 'pydevd' in str(sys.gettrace()):
+        if "PYTEST_CURRENT_TEST" in os.environ:
             logger.info('debugger detected, running without multiprocessing')
             self.multiprocessing = False
 

--- a/source/fab/cui/arguments.py
+++ b/source/fab/cui/arguments.py
@@ -98,7 +98,10 @@ class FabArgumentParser(argparse.ArgumentParser):
 
         super().__init__(*args, **kwargs)
 
-        if self.prog == "__main__.py" and kwargs.get("prog", None) is None:
+        # Python 3.14 changes the value of .prog to be
+        # "python -m pytest"
+        if (self.prog in ["python -m pytest", "__main__.py"] and
+                kwargs.get("prog", None) is None):
             # Try to pick up a better program name from the environment
             # or just use a default string
             self.prog = os.environ.get("__PROGNAME", "fab")

--- a/source/fab/tools/flags.py
+++ b/source/fab/tools/flags.py
@@ -14,9 +14,10 @@ AbstractFlags:
 AlwaysFlags(AbstractFlags):
     Flags that will always apply, independent of the path of the
     file to be compiled. It provides Python Template functionality:
-        - `$source` for *<project workspace>/source*
-        - `$output` for *<project workspace>/build_output*
-        - `$relative` for *<the source file's folder>*
+
+    - `$source` for *<project workspace>/source*
+    - `$output` for *<project workspace>/build_output*
+    - `$relative` for *<the source file's folder>*
 
 MatchFlags(AlwaysFlags)
     Flags that are only applied if a wildcard search matches the

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -14,6 +14,7 @@ from pytest import fixture
 from pytest_subprocess.fake_process import FakeProcess, ProcessRecorder
 
 from fab.build_config import BuildConfig
+from fab.tools.category import Category
 from fab.tools.compiler import CCompiler, FortranCompiler
 from fab.tools.linker import Linker
 from fab.tools.tool_box import ToolBox
@@ -194,6 +195,11 @@ def stub_tool_repository(stub_fortran_compiler,
     monkeypatch.setattr(stub_c_compiler, 'check_available', return_true)
     monkeypatch.setattr(stub_linker, 'check_available', return_true)
     tool_repository = ToolRepository()
+    # Remove all compiler and linker, since they might exist on the system
+    # running these tests, which can break tests.
+    for cat in [Category.C_COMPILER, Category.FORTRAN_COMPILER,
+                Category.LINKER]:
+        tool_repository[cat] = []
     tool_repository.add_tool(stub_fortran_compiler)
     tool_repository.add_tool(stub_c_compiler)
     tool_repository.add_tool(stub_linker)

--- a/tests/unit_tests/test_cui_arguments.py
+++ b/tests/unit_tests/test_cui_arguments.py
@@ -295,4 +295,6 @@ class TestParser:
 
         captured = capsys.readouterr()
 
+        # Python 3.14 changes the behaviour, the output starts with
+        # python -m pytest
         assert captured.out.startswith("fab ")

